### PR TITLE
histograms: Add remote-write support for Float Histograms

### PR DESCRIPTION
**Pull request by marctc**
This PR adds  remote-write support for the newly introduced float histograms(#11522).
This is was a pending feature to add after introducing float histograms.
The implementation is mostly based on this [PR](https://github.com/prometheus/prometheus/pull/11001)

### CODE REVIEW

File: storage/remote/queue_manager.go [hunk 1]
```diff
@@ -1325,6 +1380,7 @@ func (s *shards) runShard(ctx context.Context, shardID int, queue *queue) {
 			droppedSamples := int(s.enqueuedSamples.Load())
 			droppedExemplars := int(s.enqueuedExemplars.Load())
 			droppedHistograms := int(s.enqueuedHistograms.Load())
+			droppedFloatHistograms := int(s.enqueuedFloatHistograms.Load())
 			s.qm.metrics.pendingSamples.Sub(float64(droppedSamples))
 			s.qm.metrics.pendingExemplars.Sub(float64(droppedExemplars))
 			s.qm.metrics.pendingHistograms.Sub(float64(droppedHistograms))
```
Comment (LeviHarrison) [inline storage/remote/queue_manager.go:1386]: This should be done for float histograms too Same with `failedHistogramsTotal` below and these places:

* https://github.com/prometheus/prometheus/blob/87368d69d82be6846ebefe36b985a4ec542fec88/storage/remote/queue_manager.go#L1529

* https://github.com/prometheus/prometheus/blob/87368d69d82be6846ebefe36b985a4ec542fec88/storage/remote/queue_manager.go#L1544

File: storage/remote/queue_manager.go [hunk 2]
```diff
@@ -1018,9 +1065,10 @@ type shards struct {
 	qm     *QueueManager
 	queues []*queue
 	// So we can accurately track how many of each are lost during shard shutdowns.
-	enqueuedSamples    atomic.Int64
-	enqueuedExemplars  atomic.Int64
-	enqueuedHistograms atomic.Int64
+	enqueuedSamples         atomic.Int64
+	enqueuedExemplars       atomic.Int64
+	enqueuedHistograms      atomic.Int64
+	enqueuedFloatHistograms atomic.Int64
```
Comment (codesome) [inline storage/remote/queue_manager.go:1071]: Let's track all the numbers in the existing histogram metrics itself (here and elsewhere). 'float histogram' is just a different in-memory representation of native histogram while at the end they are just histograms.

File: storage/remote/queue_manager_test.go [hunk 3]
```diff
@@ -617,23 +628,56 @@ func createHistograms(numSamples, numSeries int) ([]record.RefHistogramSample, [
...
+				FH: &histogram.FloatHistogram{
+					Schema:          2,
+					ZeroThreshold:   1e-128,
+					ZeroCount:       0,
+					Count:           2,
+					Sum:             0,
+					PositiveSpans:   []histogram.Span{{Offset: 0, Length: 1}},
+					PositiveBuckets: []float64{float64(i) + 1},
+					NegativeSpans:   []histogram.Span{{Offset: 0, Length: 1}},
+					NegativeBuckets: []float64{float64(-i) - 1},
```
Comment (codesome) [inline storage/remote/queue_manager_test.go:649]: Buckets in float histograms are absolute values, so this histogram is invalid. How about we reuse the above `createHistograms` and pass a `floatHistogram` bool, and use ToFloat() where appropriate? The return can have 3 slices, of which 2 are histograms, and depending on which one you ask for, only use that slice.

File: tsdb/wlog/watcher_test.go [hunk 4]
```diff
@@ -188,6 +195,25 @@ func TestTailSamples(t *testing.T) {
...
+						FH: &histogram.FloatHistogram{
+							Schema:          2,
+							ZeroThreshold:   1e-128,
+							ZeroCount:       0,
+							Count:           2,
+							Sum:             0,
+							PositiveSpans:   []histogram.Span{{Offset: 0, Length: 1}},
+							PositiveBuckets: []float64{float64(i) + 1},
+							NegativeSpans:   []histogram.Span{{Offset: 0, Length: 1}},
+							NegativeBuckets: []float64{float64(-i) - 1},
```
Comment (codesome) [inline tsdb/wlog/watcher_test.go:212]: Buckets must be absolute values

File: storage/remote/queue_manager.go [hunk 5]
```diff
@@ -1132,6 +1181,9 @@ func (s *shards) enqueue(ref chunks.HeadSeriesRef, data timeSeries) bool {
 		case tHistogram:
 			s.qm.metrics.pendingHistograms.Inc()
 			s.enqueuedHistograms.Inc()
+		case tFloatHistogram:
+			s.qm.metrics.pendingHistograms.Inc()
+			s.enqueuedHistograms.Inc()
```
Comment (codesome) [inline storage/remote/queue_manager.go:1186]: You can club this with 
```
case tHistogram, tFloatHistogram:
	s.qm.metrics.pendingHistograms.Inc()
	s.enqueuedHistograms.Inc()
```

File: storage/remote/queue_manager.go [hunk 6]
```diff
@@ -1032,10 +1079,11 @@ type shards struct {
...
 	// after giving them a chance to terminate.
-	hardShutdown                    context.CancelFunc
-	samplesDroppedOnHardShutdown    atomic.Uint32
-	exemplarsDroppedOnHardShutdown  atomic.Uint32
-	histogramsDroppedOnHardShutdown atomic.Uint32
+	hardShutdown                         context.CancelFunc
+	samplesDroppedOnHardShutdown         atomic.Uint32
+	exemplarsDroppedOnHardShutdown       atomic.Uint32
+	histogramsDroppedOnHardShutdown      atomic.Uint32
+	floatHistogramsDroppedOnHardShutdown atomic.Uint32
```
Comment (codesome) [inline storage/remote/queue_manager.go:1086]: I think this is no longer required and is not being used anywhere

File: tsdb/wlog/watcher_test.go [hunk 7]
```diff
@@ -171,22 +177,31 @@ func TestTailSamples(t *testing.T) {
...
-							PositiveSpans:   []histogram.Span{{Offset: 0, Length: 1}},
-							PositiveBuckets: []int64{int64(i) + 1},
-							NegativeSpans:   []histogram.Span{{Offset: 0, Length: 1}},
-							NegativeBuckets: []int64{int64(-i) - 1},
-						},
+						H:   hist,
 					}}, nil)
+
 					require.NoError(t, w.Log(histogram))
+					floatHistogram := enc.FloatHistogramSamples([]record.RefFloatHistogramSample{{
```
Comment (codesome) [inline tsdb/wlog/watcher_test.go:199]: We are not checking if we watched the float histograms in this test. There is `expectedHistograms` for integer histograms; we need one for float histograms.

**Review by codesome — COMMENTED** [review]
Nice work! We can merge after fixing these comments.

**Pull request merged**